Rework test layers to use cmake generators

lenny-lunarg opened the pull request:
This PR uses CMake generators to generate the manifest json files used by the test layers. This means that we no longer need to keep a json file for each supported platform, and there's also no need to manually update the API version stored in the manifest files. Both of these fields are filled in by CMake a compile time.

karl-lunarg commented on tests/layers/CMakeLists.txt:
```diff
@@ -114,3 +68,33 @@ add_vk_layer(wrap_objects ${WRAP_SRCS})
 set(TEST_SRCS test.cpp vk_layer_table.cpp vk_layer_extension_utils.cpp)
 add_vk_layer(test ${TEST_SRCS})
+
+if(WIN32 OR CMAKE_GENERATOR MATCHES "^Xcode.*")
```
You're really trying to determine if you have a multi-config generator here, and this test mixes up platform and generator.  Also, I'm not sure if there isn't a way to do a WIN32/Windows build on something other than Visual Studio and if that way uses a single-config or multi-config generator.

It may be better to check to see if the CMAKE_CONFIGURATION_TYPES list is empty or not, where not empty means multi-config and that this path must be taken.

Note that there is a new GENERATOR_IS_MULTI_CONFIG property that is more reliable, but it does not show up until CMake 3.9.

I do recognize that CMake code elsewhere in this repo and other repos that use the approach used here or something like it, but perhaps we could start migrating to something better in new code.

And if checking CMAKE_CONFIGURATION_TYPES is too uncomfortable, then maybe checking for MSVC instead of WIN32 might be better.

lenny-lunarg replied:
I wrote this as a way of simplifying the old logic, but you're right that checking for Windows or XCode isn't the cleanest solution. The GENERATOR_IS_MULTI_CONFIG property sounds perfect here, except for the version restrictions. But I do think your solution of checking if its Visual Studio or XCode at least sounds better than the logic I currently have. I'll toy around with it and see what seems to work best.

karl-lunarg commented on tests/layers/CMakeLists.txt:
```diff
@@ -114,3 +68,33 @@ add_vk_layer(wrap_objects ${WRAP_SRCS})
+    set(JSON_DIR "${CMAKE_CURRENT_BINARY_DIR}")
+endif()
+
+# Run each .json.in fule through the generator
```
This is semi-tricky.  Might want to add a comment that it is necessary to create a Cmake command file so that it can run during the build phase.

karl-lunarg commented on tests/layers/CMakeLists.txt:
```diff
@@ -114,3 +68,33 @@ add_vk_layer(wrap_objects ${WRAP_SRCS})
+endforeach()
+
+# On some platforms we may also need to generator the directory
+if(WIN32 OR CMAKE_GENERATOR MATCHES "^Xcode.*")
```
same comment as above

karl-lunarg commented on tests/layers/CMakeLists.txt:
```diff
@@ -114,3 +68,33 @@ add_vk_layer(wrap_objects ${WRAP_SRCS})
+    add_custom_target(${CONFIG_FILE}-json ALL COMMAND ${CMAKE_COMMAND} ${CONFIG_DEFINES} -P "${CMAKE_CURRENT_BINARY_DIR}/generator.cmake")
+endforeach()
+
+# On some platforms we may also need to generator the directory
```
_create_ the directory?

mikew-lunarg commented:
@mark-lunarg I'd be glad to consider that after my next wave of CMake rework.  I'd rather limit perturbations while WIP.

lenny-lunarg commented:
After reading the feedback from @karl-lunarg, I reworked a couple of things.

1. I no longer do the check for WIN32 or XCode, nor do I use GENERATOR_IS_MULTI_CONFIG or anything similar. Instead, I use a cmake generator expression to determine the output directory of the layer target, and I just write the json files to that directory.

2. I added a bunch of comments to the `tests/layers/CmakeLists.txt` file. I think this covers why its doing what its doing, but let me know if anyone disagrees.

3. I found that the build step to make directories was unneeded. The old method where we copied or created a link needed it, but the CMake `configure_file` command will create directories as needed.

4. I discovered that Windows tests were actually silently failing in my last iteration. The cause was that I was using a forward slash in the json file. I added some code to determine platform specific paths in the CMake file and use that properly.

lenny-lunarg commented on tests/layers/CMakeLists.txt:
```diff
@@ -114,3 +68,32 @@ add_vk_layer(wrap_objects ${WRAP_SRCS})
+    # The output file needs Unix "/" separators or Windows "\" separators
+    # On top of that, Windows separators actually need to be doubled because the json format uses backslash escapes
+    file(TO_NATIVE_PATH "./" RELATIVE_PATH_PREFIX)
+    string(REPLACE "\\" "\\\\" RELATIVE_PATH_PREFIX "${RELATIVE_PATH_PREFIX}")
```
Good point I'll move them.

lenny-lunarg commented on tests/layers/CMakeLists.txt:
```diff
@@ -114,3 +68,32 @@ add_vk_layer(wrap_objects ${WRAP_SRCS})
+# We need to create the generator.cmake script so that the generator can be run at compile time, instead of configure time
+# Running at compile time lets us use cmake generator expressions (TARGET_FILE_NAME and TARGET_FILE_DIR, specifically)
+file(WRITE "${CMAKE_CURRENT_BINARY_DIR}/generator.cmake" "configure_file(\"\${INPUT_FILE}\" \"\${OUTPUT_FILE}\")")
+foreach(CONFIG_FILE ${LAYER_JSON_FILES})
```
I just used CONFIG_FILE because that was the old name, but you're right that it could be improved. I'd be tempted to use it BINARY_TARGET_NAME(S) as names, since that usually is the name of the layer binary's target, but that doesn't apply very well to metalayers.

I'm not sure what is best.

karl-lunarg replied:
Maybe just TARGET_NAMES.   But if none of these is obviously good, just stay with CONFIG_FILE and don't delay cuz of this.

lenny-lunarg replied:
I went with TARGET_NAME(S). That's the best name I've seen here, so I'm going with that.

MarkY-LunarG commented:
FYI, I've create a change in PR #56 where I now automate the detection of the Vulkan versions (major/minor/patch) from the header file.  I think these two changes go hand-in-hand, although I don't think we should block one for the other.

karl-lunarg commented:
Whichever one lands last should probably replace the code in the top-level CMake file that looks up the header version with the variable set by the change in #56.
